File: RUST/src/dfs.rs
use std::collections::VecDeque;

// Define a position in the grid as a tuple (row, col)
type Position = (usize, usize);

fn get_index(row: usize, col: usize, cols: usize) -> usize {
    row * cols + col
}

fn is_wall(grid: &Vec<i32>, row: usize, col: usize, cols: usize) -> bool {
    let index = get_index(row, col, cols);
    grid[index] == 1
}

// Function to check if a position is valid and not visited
fn is_valid(row: isize, col: isize, rows: usize, cols: usize, visited: &Vec<Vec<bool>>) -> bool {
    // If cell is out of bounds
    if row < 0 || col < 0 || row >= rows as isize || col >= cols as isize {
        return false;
    }

    // If the cell is already visited
    if visited[row as usize][col as usize] {
        return false;
    }

    // Otherwise, it can be visited
    true
}

fn dfs(
    grid: &mut Vec<i32>,
    start: Position,
    end: Position,
    rows: usize,
    cols: usize,
) -> (Option<Vec<Position>>, Vec<usize>) {
    // Initialize visited array
    let mut visited = vec![vec![false; cols]; rows];

    // Initialize parent map for path reconstruction
    let mut parent: Vec<Vec<Option<Position>>> = vec![vec![None; cols]; rows];

    // Track visited nodes in order
    let mut visited_order = Vec::new();

    // Initialize direction vectors (up, down, left, right) - changed order to prioritize up-down
    let dx: [isize; 4] = [0, 0, -1, 1];
    let dy: [isize; 4] = [-1, 1, 0, 0];

    // Initialize a stack for DFS
    let mut stack = Vec::new();
    stack.push(start);

    // Flag to track if end was found
    let mut found_end = false;

    // Perform DFS
    while !stack.is_empty() {
        // Pop the top position
        let current = stack.pop().unwrap();
        let (row, col) = current;

        // Check if the current cell is valid
        if visited[row][col] {
            continue;
        }

        // Mark as visited
        visited[row][col] = true;
        visited_order.push(get_index(row, col, cols));

        // Check if we've reached the end
        if current == end {
            found_end = true;
            break;
        }

        // Push all adjacent cells in reverse order (to match the exploration order)
        for i in (0..4).rev() {
            let adj_row = row as isize + dy[i];
            let adj_col = col as isize + dx[i];

            if is_valid(adj_row, adj_col, rows, cols, &visited) {
                let adj_row = adj_row as usize;
                let adj_col = adj_col as usize;

                // Skip walls
                if is_wall(grid, adj_row, adj_col, cols) {
                    continue;
                }

                // Set parent for path reconstruction
                parent[adj_row][adj_col] = Some(current);

                // Push to stack
                stack.push((adj_row, adj_col));
            }
        }
    }

    // Reconstruct path if end was found
    let path = if found_end {
        let mut path = Vec::new();
        let mut current = end;

        // Work backwards from end to start
        while current != start {
            path.push(current);

            if let Some(prev) = parent[current.0][current.1] {
                current = prev;
            } else {
                break; // Should not happen if end was found
            }
        }

        // Add start position
        path.push(start);

        // Reverse to get path from start to end
        path.reverse();

        // Mark path cells in the grid
        for &(row, col) in &path {
            let index = get_index(row, col, cols);
            if grid[index] != 1 { // Don't mark walls
                grid[index] = 3; // Mark as path
            }
        }

        Some(path)
    } else {
        None
    };

    // Mark all visited cells in the grid (if they're not already part of the path)
    for &index in &visited_order {
        if grid[index] == 0 {
            grid[index] = 2; // Mark as visited
        }
    }

    (path, visited_order)
}

// Get the indexes of nodes in path
fn get_path_indexes(path: &Vec<Position>, cols: usize) -> Vec<usize> {
    path.iter()
        .map(|&(row, col)| get_index(row, col, cols))
        .collect()
}

pub fn find_shortest_path(
    mut grid: Vec<i32>,
    start_row: usize,
    start_col: usize,
    end_row: usize,
    end_col: usize,
    rows: usize,
    cols: usize,
) -> (Vec<i32>, Vec<usize>, Vec<usize>) {
    let start = (start_row, start_col);
    let end = (end_row, end_col);

    let (path_opt, visited_order) = dfs(&mut grid, start, end, rows, cols);

    // Get path indexes if a path was found
    let path_indexes = match path_opt {
        Some(path) => get_path_indexes(&path, cols),
        None => Vec::new()
    };

    (grid, visited_order, path_indexes)
}

File: Algorithims/DFS.cpp
#include <vector>
#include <stack>
#include <tuple>
#include <utility>

using namespace std;

using Position = pair<size_t, size_t>;

size_t get_index(size_t row, size_t col, size_t cols) {
    return row * cols + col;
}

bool is_wall(const vector<int>& grid, size_t row, size_t col, size_t cols) {
    return grid[get_index(row, col, cols)] == 1;
}

bool is_valid(int row, int col, size_t rows, size_t cols, const vector<vector<bool>>& visited) {
    if (row < 0 || col < 0 || row >= (int)rows || col >= (int)cols)
        return false;
    if (visited[row][col])
        return false;
    return true;
}

tuple<vector<int>, vector<size_t>, vector<size_t>> find_shortest_path(
    vector<int> grid,
    size_t start_row,
    size_t start_col,
    size_t end_row,
    size_t end_col,
    size_t rows,
    size_t cols
) {
    Position start = {start_row, start_col};
    Position end = {end_row, end_col};

    vector<vector<bool>> visited(rows, vector<bool>(cols, false));
    vector<vector<optional<Position>>> parent(rows, vector<optional<Position>>(cols, nullopt));
    vector<size_t> visited_order;

    int dx[4] = {0, 0, -1, 1};
    int dy[4] = {-1, 1, 0, 0};

    stack<Position> stk;
    stk.push(start);
    bool found_end = false;

    while (!stk.empty()) {
        auto [row, col] = stk.top();
        stk.pop();

        if (visited[row][col])
            continue;

        visited[row][col] = true;
        visited_order.push_back(get_index(row, col, cols));

        if (make_pair(row, col) == end) {
            found_end = true;
            break;
        }

        for (int i = 3; i >= 0; --i) {
            int adj_row = (int)row + dy[i];
            int adj_col = (int)col + dx[i];

            if (is_valid(adj_row, adj_col, rows, cols, visited)) {
                size_t arow = adj_row, acol = adj_col;
                if (is_wall(grid, arow, acol, cols))
                    continue;

                parent[arow][acol] = {row, col};
                stk.push({arow, acol});
            }
        }
    }

    vector<Position> path;
    if (found_end) {
        Position current = end;
        while (current != start) {
            path.push_back(current);
            if (parent[current.first][current.second])
                current = *parent[current.first][current.second];
            else
                break;
        }
        path.push_back(start);
        reverse(path.begin(), path.end());

        for (const auto& [r, c] : path) {
            size_t idx = get_index(r, c, cols);
            if (grid[idx] != 1)
                grid[idx] = 3;
        }
    }

    for (size_t idx : visited_order) {
        if (grid[idx] == 0)
            grid[idx] = 2;
    }

    vector<size_t> path_indexes;
    for (const auto& [r, c] : path) {
        path_indexes.push_back(get_index(r, c, cols));
    }

    return {grid, visited_order, path_indexes};
}
